lib/synapse/service_watcher/asg.rb

```diff
@@ -0,0 +1,139 @@
+require 'synapse/service_watcher/base'
+require 'aws-sdk'
+
+class Synapse::ServiceWatcher
+  class AsgWatcher < BaseWatcher
+
+    attr_reader :check_interval
+
+    def start
+      region = @discovery['aws_region'] || ENV['AWS_REGION']
+      log.info "Connecting to EC2 region: #{region}"
+
+      @asg_client = AWS::AutoScaling::Client.new(
+          region:            region,
+          access_key_id:     @discovery['aws_access_key_id']     || ENV['AWS_ACCESS_KEY_ID'],
+          secret_access_key: @discovery['aws_secret_access_key'] || ENV['AWS_SECRET_ACCESS_KEY'] )
+
+      @ec2_client = AWS::EC2::Client.new(
+          region:            region,
+          access_key_id:     @discovery['aws_access_key_id']     || ENV['AWS_ACCESS_KEY_ID'],
+          secret_access_key: @discovery['aws_secret_access_key'] || ENV['AWS_SECRET_ACCESS_KEY'] )
+
+      @check_interval = @discovery['check_interval'] || 15.0
+
+      log.info "synapse: asg watcher looking for instances in asg #{@discovery["asg_name"]}"
+
+      @watcher = Thread.new { watch }
+    end
+
+    private
+
+    def validate_discovery_opts
+      # Required, via options only.
+      raise ArgumentError, "invalid discovery method #{@discovery['method']}" \
+        unless @discovery['method'] == 'asg'
+      raise ArgumentError, "aws asg name is required for service #{@name}" \
+        unless @discovery['asg_name']
+
+      # As we're only looking up instances with hostnames/IPs, need to
+      # be explicitly told which port the service we're balancing for listens on.
+      unless @haproxy['server_port_override']
+        raise ArgumentError,
+              "Missing server_port_override for service #{@name} - which port are backends listening on?"
+      end
+
+      unless @haproxy['server_port_override'].to_s.match(/^\d+$/)
+        raise ArgumentError, "Invalid server_port_override value"
+      end
+
+      # aws region is optional in the SDK, aws will use a default value if not provided
+      unless @discovery['aws_region'] || ENV['AWS_REGION']
+        log.info "aws region is missing, will use default"
+      end
+      # access key id & secret are optional, might be using IAM instance profile for credentials
+      unless ((@discovery['aws_access_key_id'] || ENV['aws_access_key_id']) \
+              && (@discovery['aws_secret_access_key'] || ENV['aws_secret_access_key'] ))
+        log.info "aws access key id & secret not set in config or env variables for service #{name}, will attempt to use IAM instance profile"
+      end
+    end
+
+    def watch
+      until @should_exit
+        begin
+          start = Time.now
+          if set_backends(discover_instances)
+            log.info "synapse: asg watcher backends have changed."
+          end
+        rescue Exception => e
+          log.warn "synapse: error in asg watcher thread: #{e.inspect}"
+          log.warn e.backtrace
+        ensure
+          sleep_until_next_check(start)
```

Comment:
Polling is definitely ok as a V1 or for small infras, but what are your thoughts on subscribing to a SNS topic of [autoscaling notifications](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/ASGettingNotifications.html). It could be something super simple like "if we get any SNS notification at all, rediscover backends"